### PR TITLE
Add LodeRunner test to ngsa-app C#  CI/CD on Pull Request

### PR DESCRIPTION
**Pull request by gortegaMS**
Added LodeRunner test to ngsa-app C#  CI/CD on Pull Request

Partially Closes: [Issue #230](https://github.com/retaildevcrews/wcnp/issues/230)

Reference: [ngsa-java PR #24](https://github.com/retaildevcrews/ngsa-java/pull/24)

### CODE REVIEW

File: .github/workflows/ngsa-pr.yaml [hunk 1]
```diff
@@ -27,16 +27,45 @@ jobs:
...
+            echo -n ${{ secrets.NGSA_PRE_COSMOS_RO_KEY}} >| secrets/CosmosKey
+            echo -n ${{ secrets.NGSA_PRE_COSMOS_URL }} >| secrets/CosmosUrl
     - name: Docker Build
       run: |
-        docker build . -t image
-        
+        docker build . --progress auto -t image
+
+    - name: Docker Run and Test
+      run: |
```
Comment (bartr) [inline .github/workflows/ngsa-pr.yaml:39]: why are these comments here?

File: .github/workflows/ngsa-pr.yaml [hunk 2]
```diff
@@ -27,16 +27,45 @@ jobs:
         registry: ghcr.io
         username: ${{ secrets.GHCR_ID }}
         password: ${{ secrets.GHCR_PAT }}
-
+    - name: Set Secrets
+      run: |
+            echo -n ${{ secrets.NGSA_PRE_COSMOS_RO_KEY}} >| secrets/CosmosKey
```
Comment (bartr) [inline .github/workflows/ngsa-pr.yaml:32]: you just put the secret in the docker image. If you publish the image, you have a security hole.

The secrets need to be copied to a local directory and mounted with -v

File: .github/workflows/ngsa-pr.yaml [hunk 3]
```diff
@@ -27,16 +27,45 @@ jobs:
...
+
+    - name: Docker Run and Test
+      run: |
+            #docker run --rm image --version
+            #docker network create runandtestnet 
+            #docker run --rm --net=runandtestnet --network-alias=ngsacsapi --name ngsacs -p 8080:8080 image
+            
+            docker run --rm -d --name ngsacs -p 8080:8080 image
+
+            echo "Waiting for web server to start ..."
```
Comment (bartr) [inline .github/workflows/ngsa-pr.yaml:46]: why is the sleep hard-coded? In the previous version we used a loop

File: .github/workflows/ngsa-pr.yaml [hunk 4]
```diff
@@ -27,16 +27,45 @@ jobs:
...
+            sleep 25
+
+            #docker run --rm --net=runandtestnet ghcr.io/retaildevcrews/ngsa-lr:beta -l 500 -s http://ngsacsapi:8080/ --max-errors 1 -f baseline.json 
+
```
Comment (bartr) [inline .github/workflows/ngsa-pr.yaml:50]: why not use WebV?

File: .github/workflows/ngsa-pr.yaml [hunk 5]
```diff
@@ -27,16 +27,45 @@ jobs:
...
+            docker run --rm --network=host ghcr.io/retaildevcrews/ngsa-lr:beta -s "http://localhost:8080" --max-errors 1 -f baseline.json
+
+    - name: Display coverage results
```
Comment (bartr) [inline .github/workflows/ngsa-pr.yaml:53]: you won't run in test mode, so you won't get any coverage results

File: .github/workflows/ngsa-pr.yaml [hunk 6]
```diff
@@ -27,16 +27,45 @@ jobs:
...
+    - name: Delete Secrets
+      run: |
+            rm -rf ./secrets
+
     - name: Docker Tag and Push
       run: |
 
-        # run image with --version flag
-        docker run --rm image --version
+        #VERSION=$(docker run --rm image --version)
```
Comment (bartr) [inline .github/workflows/ngsa-pr.yaml:68]: you just pushed your CosmosKey to a public container registry ...

File: .github/workflows/ngsa-pr.yaml [hunk 7]
```diff
@@ -27,16 +27,45 @@ jobs:
...
+
+    - name: Display coverage results
+      run: |
+           # wait for ngsa-lr to shutdown
+           sleep 15
+
+           # display the results
+           docker logs ngsacs | tail -n 24
+    
+    - name: Delete Secrets
```
Comment (bartr) [inline .github/workflows/ngsa-pr.yaml:61]: why is this here? It doesn't do what you want it to do.

File: .github/workflows/ngsa-pr.yaml (hunk 1 above)
Comment (gortegaMS) [reply .github/workflows/ngsa-pr.yaml:39]: comments removed

File: .github/workflows/ngsa-pr.yaml (hunk 2 above)
Comment (gortegaMS) [reply .github/workflows/ngsa-pr.yaml:32]: security hole fixed, run ngsa-app container with in-memory flag instead.

File: .github/workflows/ngsa-pr.yaml (hunk 3 above)
Comment (gortegaMS) [reply .github/workflows/ngsa-pr.yaml:46]: Utilized While loop logic to wait 10 secs, then wait for 1 sec up to 30 secs then time out

File: .github/workflows/ngsa-pr.yaml (hunk 5 above)
Comment (gortegaMS) [reply .github/workflows/ngsa-pr.yaml:53]: removed coverage results check , was not needed.

File: .github/workflows/ngsa-pr.yaml (hunk 6 above)
Comment (gortegaMS) [reply .github/workflows/ngsa-pr.yaml:68]: security hole fixed, run ngsa-app container with in-memory flag instead. not secrets utilized.

image was pushed to my private container registry but I deleted it already.

File: .github/workflows/ngsa-pr.yaml (hunk 7 above)
Comment (gortegaMS) [reply .github/workflows/ngsa-pr.yaml:61]: After a second thought this is not needed since we do not have coverage results

File: .github/workflows/ngsa-pr.yaml (hunk 4 above)
Comment (gortegaMS) [reply .github/workflows/ngsa-pr.yaml:50]: After a discussion with Joseph he explained that WebV would be easier and faster to run, and it is a global dotnet tool, however we would be limited to add specific capabilities to it if we need to, on the other hand Loderunner might be better long term if we find ourselves doing so in the feature.

File: .github/workflows/ngsa.yaml [hunk 8]
```diff
@@ -35,15 +35,40 @@ jobs:
...
+                if [ $wait_time -gt 30 ] ; then
+                  echo -e "\n timeout waiting for web server to start"
+                  exit 1
+                fi
+                
+                sleep 1
+                ((wait_time=wait_time+1))
+            done
+
+            docker run --rm --network=host ghcr.io/retaildevcrews/ngsa-lr:beta -s "http://localhost:8080" --max-errors 1 -f baseline.json
```
Comment (jofultz) [inline .github/workflows/ngsa.yaml:66]: I have the same concern on Java that we need to rethink running the test on push.  I've posted a note that I'll convert to an issue.  We need to consider if we should run the test at all, run it always, or run it only when triggered by the cron job.

**Review by jofultz — APPROVED** [review]
